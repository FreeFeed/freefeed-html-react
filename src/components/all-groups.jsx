/* global CONFIG */
import { memo, useEffect, useMemo, useState, useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { sortBy, range } from 'lodash';
import { Link } from 'react-router';
import { Helmet } from 'react-helmet';

import { faCaretDown, faUserFriends } from '@fortawesome/free-solid-svg-icons';
import { getAllGroups } from '../redux/action-creators';
import { Icon } from './fontawesome-icons';

import UserName from './user-name';
import TimeDisplay from './time-display';

import styles from './all-groups.module.scss';
import { UserPicture } from './user-picture';
import { HorScrollable } from './hor-scrollable';

export default function AllGroups() {
  const dispatch = useDispatch();
  const status = useSelector((state) => state.allGroupsStatus);

  useEffect(() => void (status.success || status.loading || dispatch(getAllGroups())), [
    dispatch,
    status,
  ]);

  return (
    <div className="content">
      <Helmet title={`All Groups - ${CONFIG.siteTitle}`} defer={false} />
      <div className="box">
        <div className="box-header-timeline" role="heading">
          All Groups
        </div>
        <div className="box-body">
          <div className="row" style={{ marginBottom: 20 }}>
            <div className="col-md-6">
              <Link to="/groups">List of your groups</Link>
            </div>
            <div className="col-md-6 text-right">
              <Link to="/groups/create">Create a group</Link>
            </div>
          </div>

          {status.loading && <p>Loading...</p>}
          {status.error && (
            <p className="alert alert-danger">Can not load groups list: {status.errorText}</p>
          )}
          {status.success && <GroupsList pageSize={50} />}
        </div>
      </div>
    </div>
  );
}

const SORT_BY_SUBSCRIBERS = 'subscribers';
const SORT_BY_POSTS = 'posts';
const SORT_BY_VARIETY = 'variety';
const SORT_BY_DATE = 'date';
const SORT_DEFAULT = SORT_BY_SUBSCRIBERS;

const sortFields = {
  [SORT_BY_POSTS]: 'postsByMonth',
  [SORT_BY_SUBSCRIBERS]: 'subscribers',
  [SORT_BY_VARIETY]: 'authorsVariety',
  [SORT_BY_DATE]: 'createdAt',
};

function GroupsList({ pageSize }) {
  const [nameFilter, setNameFilter] = useState('');

  const location = useSelector((state) => state.routing.locationBeforeTransitions);
  const { groups, withProtected } = useSelector((state) => state.allGroups);

  const nameFilterLowercase = nameFilter.toLowerCase();
  const filteredGroups = nameFilter
    ? groups.filter((g) => {
        return (
          g.username.toLowerCase().includes(nameFilterLowercase) ||
          g.screenName.toLowerCase().includes(nameFilterLowercase)
        );
      })
    : groups;

  const page = getPageNumber(location);
  const sort = getSortMode(location);
  const totalPages = Math.ceil(filteredGroups.length / pageSize);

  const groupsOnPage = useMemo(
    () =>
      sortBy(filteredGroups, sortFields[sort])
        .reverse()
        .slice(page * pageSize, (page + 1) * pageSize),
    [filteredGroups, page, pageSize, sort],
  );

  return (
    <>
      {withProtected ? (
        <p>
          There are {groups.length} public and protected groups in {CONFIG.siteTitle}. Protected
          groups can only be seen by authenticated users.
        </p>
      ) : (
        <p>
          There are {groups.length} public groups in {CONFIG.siteTitle}.
        </p>
      )}
      <p>
        <label htmlFor="groups-name-filter">Filter by username or display name</label>
        <input
          id="groups-name-filter"
          type="search"
          placeholder="Type to filter"
          onChange={useCallback((e) => setNameFilter(e.target.value), [])}
          className="form-control narrow-input"
        />
      </p>
      <p>Click on the column headers to change table sorting order.</p>

      <HorScrollable>
        <table className={styles.table}>
          <thead>
            <tr className={styles.headersRow}>
              <th className={styles.mainColumn}>Group</th>
              <SortHeader mode={SORT_BY_SUBSCRIBERS} currentMode={sort}>
                Subscribers
              </SortHeader>
              <SortHeader mode={SORT_BY_POSTS} currentMode={sort}>
                Posts per month
              </SortHeader>
              <SortHeader mode={SORT_BY_VARIETY} currentMode={sort}>
                Authors variety
              </SortHeader>
              <SortHeader mode={SORT_BY_DATE} currentMode={sort}>
                Creation date
              </SortHeader>
            </tr>
          </thead>
          <tbody>
            {groupsOnPage.map((g) => (
              <GroupRow key={g.id} g={g} />
            ))}
          </tbody>
        </table>
      </HorScrollable>
      {groupsOnPage.length === 0 ? (
        <div className={styles.empty}>No groups match current filter</div>
      ) : (
        false
      )}
      {totalPages > 1 && (
        <ul className="pagination">
          {range(1, totalPages).map((n) => (
            <li key={n} className={page === n - 1 ? 'active' : undefined}>
              <Link to={linkToPage(n, location)}>{n}</Link>
            </li>
          ))}
        </ul>
      )}
    </>
  );
}

function SortHeader({ children, mode, currentMode }) {
  const location = useSelector((state) => state.routing.locationBeforeTransitions);
  return (
    <th className={styles.sortableHeader + (mode === currentMode ? '' : ` ${styles.inactive}`)}>
      <Link to={linkToSort(mode, location)}>
        <div>
          {children}
          <Icon icon={faCaretDown} className={styles.caret} />
        </div>
      </Link>
    </th>
  );
}

const GroupRow = memo(function GroupRow({ g }) {
  const u = useSelector((state) => state.users[g.id]);
  return (
    <tr className={styles.groupRow}>
      <td>
        <UserPicture user={u} className={styles.userpic} />
        <div className={styles.groupInfo}>
          <UserName user={u}>{u.username}</UserName>
          {u.username !== u.screenName && <div className="small">{u.screenName}</div>}
          {u.isProtected === '1' && (
            <div className="muted-text small">
              <Icon icon={faUserFriends} title="Protected group" />
            </div>
          )}
        </div>
      </td>
      <td>{g.subscribers}</td>
      <td>{g.postsByMonth === 0 ? '< 1' : Math.ceil(g.postsByMonth)}</td>
      <td>{Math.round(g.authorsVariety * 100)}%</td>
      <td>
        <TimeDisplay timeStamp={+u.createdAt} absolute dateOnly />
      </td>
    </tr>
  );
});

function getPageNumber({ query: { page = '1' } }) {
  page = parseInt(page);
  if (!Number.isFinite(page) || page < 1) {
    return 0;
  }
  return page - 1;
}

function getSortMode({ query: { sort = SORT_DEFAULT } }) {
  return sortFields[sort] ? sort : SORT_DEFAULT;
}

function linkToPage(page, { pathname, query }) {
  return {
    pathname,
    query: { ...query, page: page > 1 ? page : undefined },
  };
}

function linkToSort(sort, { pathname }) {
  return {
    pathname,
    query: { sort: sort !== SORT_DEFAULT ? sort : undefined },
  };
}
