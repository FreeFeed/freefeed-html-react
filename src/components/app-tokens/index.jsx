import React from 'react';

import Tokens from './tokens';
import CreateToken from './create-token';
import CreateLink from './create-link';
import ScopesList from './scopes-list';
import Layout from './layout';


export default function ({ routeParams: { splat } }) {
  if (!splat) {
    return <Layout component={Tokens} />;
  }
  if (splat === 'create') {
    return <Layout component={CreateToken} title="Generate new token" />;
  }
  if (splat === 'create-link') {
    return <Layout component={CreateLink} title="Create magic link" />;
  }
  if (splat === 'scopes') {
    return <Layout component={ScopesList} title="Token access rights" />;
  }

  return <p>Page not found: {splat}</p>;
}

