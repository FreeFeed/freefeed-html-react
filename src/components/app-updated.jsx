/* global CONFIG */
import { useSelector } from 'react-redux';

import styles from './app-updated.module.scss';
import { ButtonLink } from './button-link';

function reloadPage() {
  window.location.reload(true);
}

export function AppUpdated() {
  const updated = useSelector((state) => state.appUpdated.updated);

  if (!updated) {
    return null;
  }

  return (
    <div className={styles.bar}>
      <div className={styles.indicator}>
        Thereâ€™s an update for {CONFIG.siteTitle}!{' '}
        <ButtonLink className={styles.refresh} onClick={reloadPage}>
          Refresh the page
        </ButtonLink>{' '}
        when you are ready.
      </div>
    </div>
  );
}
