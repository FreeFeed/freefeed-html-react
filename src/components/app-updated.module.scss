.bar {
  background-color: #337ab7;
  color: white;
}

.refresh {
  color: inherit !important;
  font-weight: bold;
  text-decoration: underline;

  &:hover {
    color: inherit !important;
  }
}

.indicator {
  margin: auto;
  padding: 0.5em 15px;

  @media (max-width: 768px) {
    max-width: 100%;
  }

  @media (min-width: 769px) and (max-width: 992px) {
    max-width: 750px;
  }
  @media (min-width: 993px) {
    max-width: 948px;
  }
}
