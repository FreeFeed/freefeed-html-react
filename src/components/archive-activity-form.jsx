import PropTypes from 'prop-types';
import React from 'react';

export default class ArchiveActivityForm extends React.Component {
  static propTypes = {
    action: PropTypes.func.isRequired,
    formState: PropTypes.shape({
      inProgress: PropTypes.bool.isRequired,
      success: PropTypes.bool.isRequired,
      error: PropTypes.bool.isRequired,
      errorText: PropTypes.string.isRequired,
    }),
  };

  state = { checked: false };

  setChecked = (e) => this.setState({ checked: e.target.checked });

  action = (e) => {
    e.preventDefault();
    if (this.canSubmit()) {
      this.props.action();
    }
  };

  canSubmit() {
    return this.state.checked && !this.props.formState.inProgress;
  }

  render() {
    const { checked } = this.state;
    const { formState } = this.props;
    return (
      <form onSubmit={this.action}>
        <div className="checkbox">
          <label>
            <input type="checkbox" checked={checked} onClick={this.setChecked} />I allow to restore
            my comments and likes of other usersâ€™ posts
          </label>
        </div>
        <div className="form-group">
          <button type="submit" className="btn btn-default" disabled={!this.canSubmit()}>
            Yes, I allow
          </button>
        </div>
        {formState.error ? (
          <div className="alert alert-danger" role="alert">
            {formState.errorText}
          </div>
        ) : (
          false
        )}
      </form>
    );
  }
}
