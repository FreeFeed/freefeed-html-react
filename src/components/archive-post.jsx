import React from 'react';
import { connect } from 'react-redux';

function ArchivePostHandler({ router, inProgress, success, id }) {
  if (success) {
    router.replace(`/archive/${encodeURIComponent(id)}`);
    return null;
  }

  let postBody = null;

  if (inProgress) {
    postBody = (
      <p>
        Loading info for <strong>{router.location.query.url}</strong>...
      </p>
    );
  } else {
    postBody = (
      <div>
        <p>
          It seems that post <strong>{router.location.query.url}</strong> has not yet been restored
          from the archive.
        </p>
        <p>
          You can try to find{' '}
          <a
            target="_blank"
            style={{ textDecoration: 'underline' }}
            href={`https://web.archive.org/web/20150410000000/${encodeURIComponent(
              router.location.query.url,
            )}`}
          >
            copy of this post on archive.org
          </a>
          .
        </p>
      </div>
    );
  }

  return (
    <div className="box">
      <div className="box-header-timeline">Friendfeed Post</div>
      <div className="box-body">{postBody}</div>
      <div className="box-footer" />
    </div>
  );
}

function selectState(state) {
  const { archivePost } = state;
  return archivePost;
}

export default connect(selectState)(ArchivePostHandler);
