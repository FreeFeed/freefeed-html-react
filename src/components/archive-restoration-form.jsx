import PropTypes from 'prop-types';
import React from 'react';
import _ from 'lodash';
import { faExternalLinkAlt } from '@fortawesome/free-solid-svg-icons';
import { Icon } from './fontawesome-icons';

export default class ArchiveRestorationForm extends React.Component {
  static propTypes = {
    action: PropTypes.func.isRequired,
    formState: PropTypes.shape({
      inProgress: PropTypes.bool.isRequired,
      success: PropTypes.bool.isRequired,
      error: PropTypes.bool.isRequired,
      errorText: PropTypes.string.isRequired,
    }),
    sources: PropTypes.array.isRequired,
    oldUsername: PropTypes.string.isRequired,
  };

  state = {
    disable_comments: false,
    via_restore: [],
  };

  setDisableComments = (e) => this.setState({ disable_comments: e.target.checked });
  setViaSource = (e) => {
    const { checked, value } = e.target;
    if (checked) {
      this.setState({ via_restore: _.union(this.state.via_restore, [value]) });
    } else {
      this.setState({ via_restore: _.without(this.state.via_restore, value) });
    }
  };
  selectAllSources = (e) => {
    if (e.target.checked) {
      this.setState({ via_restore: this.props.sources.map((s) => s.url) });
    } else {
      this.setState({ via_restore: [] });
    }
  };

  canSubmit() {
    return this.state.via_restore.length > 0 && !this.props.formState.inProgress;
  }

  action = (e) => {
    e.preventDefault();
    if (this.canSubmit()) {
      this.props.action(this.state);
    }
  };

  render() {
    const { via_restore, disable_comments } = this.state;
    const { formState, sources, oldUsername } = this.props;
    const totalCount = sources.reduce((p, s) => p + s.count, 0);
    const selectedCount = sources
      .filter((s) => via_restore.includes(s.url))
      .reduce((p, s) => p + s.count, 0);

    const buttonText =
      totalCount === selectedCount
        ? 'Restore all my posts'
        : selectedCount === 0
        ? 'Restore my posts'
        : `Restore my ${selectedCount} ${selectedCount === 1 ? 'post' : 'posts'}`;

    return (
      <form onSubmit={this.action}>
        <p>
          You have {totalCount} posts in your <strong>friendfeed.com/{oldUsername}</strong> archive.
          Posts from which source would you like to restore?
        </p>
        <div className="checkbox">
          <label>
            <input
              type="checkbox"
              checked={via_restore.length === sources.length}
              onClick={this.selectAllSources}
            />
            Restore all {totalCount} posts (or select specific sources below)
          </label>
        </div>
        <SourceList sources={sources} selected={via_restore} onClick={this.setViaSource} />
        <div className="checkbox">
          <label>
            <input type="checkbox" checked={disable_comments} onClick={this.setDisableComments} />
            Disable new comments for your posts restored from archive (you can enable comments for
            individual posts after the restoring has been completed)
          </label>
        </div>
        <div className="form-group">
          <button type="submit" className="btn btn-default" disabled={!this.canSubmit()}>
            {buttonText}
          </button>
        </div>
        {formState.error ? (
          <div className="alert alert-danger" role="alert">
            {formState.errorText}
          </div>
        ) : (
          false
        )}
        <p>
          <em>Requests are handled manually, so be prepared to wait a bit.</em>
        </p>
      </form>
    );
  }
}

const FRF_URL = 'http://friendfeed.com';

class SourceList extends React.Component {
  static propTypes = {
    sources: PropTypes.array.isRequired,
    selected: PropTypes.array.isRequired,
    onClick: PropTypes.func.isRequired,
  };

  renderSource({ name, url }) {
    if (url === FRF_URL) {
      return <span>Posts on FriendFeed</span>;
    }
    if (url.indexOf(FRF_URL) === 0) {
      return <span>FriendFeed: {name}</span>;
    }
    return (
      <span>
        {name}
        <a href={url} title={url} target="_blank" className="source-link">
          <Icon icon={faExternalLinkAlt} />
        </a>
      </span>
    );
  }

  render() {
    const { selected, onClick } = this.props;
    const sources = this.props.sources.sort((a, b) => {
      if (a.url === FRF_URL) {
        return -1;
      }
      if (b.url === FRF_URL) {
        return 1;
      }
      return b.count - a.count;
    });
    return (
      <div className="archive-source-list">
        {sources.map((s) => (
          <div className="checkbox" key={s.url}>
            <label>
              <input
                type="checkbox"
                value={s.url}
                checked={selected.includes(s.url)}
                onClick={onClick}
              />
              {this.renderSource(s)} <span className="text-muted">({s.count})</span>
            </label>
          </div>
        ))}
      </div>
    );
  }
}
