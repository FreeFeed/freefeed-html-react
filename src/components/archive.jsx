import React from 'react';
import { connect } from 'react-redux';

import {
  archiveRestoreActivity,
  archiveStartRestoration,
  resetArchiveForms,
} from '../redux/action-creators';
import ArchiveActivityForm from './archive-activity-form';
import ArchiveRestorationForm from './archive-restoration-form';

class Archive extends React.Component {
  componentWillUnmount() {
    this.props.resetArchiveForms();
  }

  render() {
    const { archives } = this.props;
    if (!archives) {
      return (
        <div className="content">
          <div className="box">
            <div className="box-header-timeline">Restore from FriendFeed.com Archives</div>
            <div className="box-body">
              <p>
                We do not have any records about your old FriendFeed account. If you think this is a
                mistake, please contact us at{' '}
                <a href="mailto:freefeed.net@gmail.com">freefeed.net@gmail.com</a>.
              </p>
            </div>
          </div>
        </div>
      );
    }
    return (
      <div className="content">
        <div className="box">
          <div className="box-header-timeline">Restore from FriendFeed.com Archives</div>
          <div className="box-body">
            <p>
              In April 2015, before FriendFeed was shut down, a group of volunteers had created an
              archive of posts for 5356 users and 1231 groups of FriendFeed.com. These archives are
              available at{' '}
              <a href="https://clio.freefeed.net/" target="_blank">
                clio.freefeed.net
              </a>
              .
            </p>
            <p>
              If you have an archive of your posts at{' '}
              <a href="https://clio.freefeed.net/" target="_blank">
                clio.freefeed.net
              </a>
              , you can now restore it to FreeFeed.
            </p>
            <p>
              Please note that all restore requests are processed manually and may take a few days
              to complete.
            </p>
            <p>
              <a href="https://dev.freefeed.net/w/archives-faq/" target="_blank">
                Restore from archives FAQ
              </a>{' '}
              (in Russian)
            </p>
            <h3>Restore your comments and likes</h3>
            {this.renderActivityForm()}
            <h3>Restore your posts</h3>
            {this.renderRestorationForm()}
          </div>
        </div>
      </div>
    );
  }

  renderActivityForm() {
    const { archives, archiveRestoreActivity, archiveActivityFormState } = this.props;
    if (archives && !archives.restore_comments_and_likes) {
      return (
        <ArchiveActivityForm action={archiveRestoreActivity} formState={archiveActivityFormState} />
      );
    }
    return <p>You allowed to restore your comments and likes of other usersâ€™ posts.</p>;
  }

  renderRestorationForm() {
    const { archives, archiveStartRestoration, archiveRestorationFormState } = this.props;

    if (!archives.has_archive) {
      return (
        <p>
          We do not have archive of your FriendFeed account (<code>{archives.old_username}</code>).
          If you think this is a mistake, please contact us at{' '}
          <a href="mailto:freefeed.net@gmail.com">freefeed.net@gmail.com</a>.
        </p>
      );
    }

    if (archives.recovery_status === 0) {
      return (
        <ArchiveRestorationForm
          sources={archives.via_sources}
          oldUsername={archives.old_username}
          action={archiveStartRestoration}
          formState={archiveRestorationFormState}
        />
      );
    }

    if (archives.recovery_status === 1) {
      return (
        <p>
          Your archive restoration is in progress. Requests are processed manually and may take a
          few days to complete.
        </p>
      );
    }

    return <p>Your archive has been restored. Thank you!</p>;
  }
}

function mapStateToProps(state) {
  return {
    user: state.user,
    archives: state.user.privateMeta.archives,
    archiveActivityFormState: state.archiveActivityForm,
    archiveRestorationFormState: state.archiveRestorationForm,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    archiveStartRestoration: (params) => dispatch(archiveStartRestoration(params)),
    archiveRestoreActivity: () => dispatch(archiveRestoreActivity()),
    resetArchiveForms: () => dispatch(resetArchiveForms()),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Archive);
