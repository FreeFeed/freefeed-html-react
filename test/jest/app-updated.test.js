/* global jest, describe, it, expect */
import { render } from '@testing-library/react';
import * as reactRedux from 'react-redux';

import { AppUpdated } from '../../src/components/app-updated';

const defaultState = {
  appUpdated: {
    updated: true,
  },
};

describe('AppUpdated', () => {
  it('Renders if there is an update', () => {
    const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
    useSelectorMock.mockImplementation((selector) => selector(defaultState));

    const { asFragment } = render(<AppUpdated />);
    expect(asFragment()).toMatchSnapshot();
  });
});
