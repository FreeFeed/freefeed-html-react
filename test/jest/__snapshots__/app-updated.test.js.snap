// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppUpdated Renders if there is an update 1`] = `
<DocumentFragment>
  <div
    class="bar"
  >
    <div
      class="indicator"
    >
      Thereâ€™s an update for FreeFeed! 
      <a
        aria-disabled="false"
        class="refresh"
        role="button"
        tabindex="0"
      >
        Refresh the page
      </a>
       when you are ready.
    </div>
  </div>
</DocumentFragment>
`;
